Clean up in string solver

romainbrenguier opened the pull request:
This deduplicate the use of `array_pool.find` and `char_array_of_pointer` and move the code for instantiation. This is preparation to make coming changes to the solver easier.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Remove unused string declaration

romainbrenguier committed:
Deduplicate char_array_of_pointer

This an alias for pool.find. It is better to keep only one to avoid
confusion.

romainbrenguier committed:
Rename find_qvar to contains

`qvar` is not relevant outside the context of the function and "find"
can mislead into thinking the function returns an iterator.

romainbrenguier committed:
Move constraint instantiation to appropriate file

string_constraint_instantiation.cpp is the appropriate place for the
instanciation functions.

tautschnig commented on src/solvers/strings/array_pool.h:
```diff
@@ -99,11 +99,6 @@ array_string_exprt of_argument(array_poolt &array_pool, const exprt &arg);
 
 /// Return the array associated to the given pointer or creates a new one
 DEPRECATED("use pool.find(pointer, length) instead")
```
Shouldn't the above also be removed?

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+#include <util/expr_iterator.h>
 #include "string_constraint_instantiation.h"
 
+/// look for the symbol and return true if it is found
```
How about: "Look for symbol \p qvar in the expression \p index and return true if it was found."

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+/// look for the symbol and return true if it is found
+/// \param index: an index expression
+/// \param qvar: a symbol expression
+/// \return a Boolean
```
"True, iff \p qvar appears in \p index."

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
 #include "string_constraint_instantiation.h"
 
+/// look for the symbol and return true if it is found
+/// \param index: an index expression
```
Does it actually matter whether this is an "index" expression?!

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+    else
+    {
+      if(positive)
+        elems[cur] = elems[cur] + 1;
```
`++elems[cur];`

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+      if(positive)
+        elems[cur] = elems[cur] + 1;
+      else
+        elems[cur] = elems[cur] - 1;
```
`--elems[cur];` (this actually makes a substantial difference here, because it avoids a duplicate map lookup)

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+
+std::map<exprt, int> map_representation_of_sum(const exprt &f)
+{
+  // number of time the leaf should be added (can be negative)
```
s/time/times/

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+sum_over_map(std::map<exprt, int> &m, const typet &type, bool negated)
+{
+  exprt sum = nil_exprt();
+  mp_integer constants = 0;
```
Don't declare or create them earlier than necessary. Notably they won't be needed at all if `m.empty()`.

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+
+  for(const auto &term : m)
+  {
+    // We should group constants together...
```
What is this comment trying to suggest?

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+    // We should group constants together...
+    const exprt &t = term.first;
+    int second = negated ? (-term.second) : term.second;
+    if(t.id() == ID_constant)
```
I suppose this is what the comment refers to - why isn't this already done in `map_representation_of_sum`?

romainbrenguier replied:
`map_representation_of_sum` just maps expressions to the number of time they appear in the sum, without doing anything special for constants, I clarified that in the example for `map_representation_of_sum`

tautschnig replied:
I understand that - but is that really the most appropriate design?

romainbrenguier replied:
Another possibility would be to use an exprt equal to 1 in the map and add all constants to its factor. The current implementation of `sum_over_map` handles both cases so I wouldn't change it. We could make `map_representation_of_sum` do more work but I don't think there is not a huge need for it.

tautschnig replied:
My interpretation of this code is that you are computing the coefficients of a polynomial. (I do not have a sufficient understanding of the code to tell why you need to do this, but at least that's my reading of this code.) If this interpretation is correct, then I'd consider renaming the function to make this clear. And if you're doing this, then certainly constants should be folded to become the coefficient of the zero-degree summand.

romainbrenguier replied:
It's actually a linear function rather than a polynomial function, and computing this representation is useful to inverse the function. I've put this in a class `linear_functiont` to make that clearer (constants are now stored in a single field).

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+            else
+              plus_exprt(sum, t);
+            for(int i = 1; i < second; i++)
+              sum = plus_exprt(sum, t);
```
Why doesn't this use multiplication?

romainbrenguier replied:
I think the original intent was to have only plus_exprt and minus_exprt but this shouldn't matter for our purpose, I will replace it by multiplication to make the code simpler

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+compute_inverse_function(const exprt &qvar, const exprt &val, const exprt &f)
+{
+  exprt positive, negative;
+  // number of time the element should be added (can be negative)
```
s/time/times/

tautschnig commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -9,8 +9,234 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
+      it->second == 0,
+      string_refinement_invariantt(
+        "a proper function must have exactly one "
+        "occurrences after reduction, or it cancelled out, and it does not"
```
s/occurrences/occurrence/

romainbrenguier committed:
Update documentation of contains

Previous version was not perfectly clear

romainbrenguier committed:
Improve map_representation_of_sum code

Fix typo, avoid double lookup and improve example in the documentation.

romainbrenguier committed:
Improve sum_over_map code

Declare variables after a return that does not use it.
Rename `second` to `factor` which reflects better what it represents.
Improve comment about the constant case.

romainbrenguier committed:
Fix sum update in sum_over_map

The instruction was without an effect and should have updated sum instead.

romainbrenguier committed:
Use multiplication in sum_over_map for factor > 1

When the factor is not -1, 0 or 1 using multiplication makes the
resulting expression simpler.

romainbrenguier committed:
Fix typos in compute_inverse_function

romainbrenguier commented:
@tautschnig I implemented your suggestions in separate commits to keep the moving commit clean.

romainbrenguier committed:
Apply clang-format suggestions in instantiate

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9e50b3f93fe9ebc0d5a4955bb2cab0c1a64897b2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108807806
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

romainbrenguier committed:
Refactor map_representation_of_sum/sum_over_map

We wrap these two functions in a `linear_functiont` and rename them, to
make clearer how these two functions work together and clarifies what
this intermediary structure represents.

romainbrenguier committed:
Refactor compute_inverse_function

We replace this function by a `solve` function which manipulate a
linear_functiont. This makes the interface and the code clearer.

romainbrenguier committed:
Make fields of linear_functiont private

This improves encapsulation.

romainbrenguier committed:
Add a linear_functiont::format for debugging

This can be useful for debugging or testing the class.

romainbrenguier committed:
Add an linear_functiont::add for linear function

This will be useful for instance to compute the inverse of a function.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 3b194e0adfe7ad32f0827f1335239b7382bbe7ed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/108870055
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

peterschrammel commented on src/solvers/strings/string_constraint_instantiation.cpp:
```diff
@@ -10,6 +10,232 @@ Author: Jesse Sigal, jesse.sigal@diffblue.com
 /// Defines related function for string constraints.
 
 #include "string_constraint_instantiation.h"
+#include <algorithm>
```
typo in commit message: instantiation

peterschrammel commented on src/solvers/strings/string_constraint_instantiation.h:
```diff
@@ -35,7 +35,7 @@ std::vector<exprt> instantiate_not_contains(
     &witnesses);
 
-/// Connonical representation of linear function, for instance, expression
+/// Cannonical representation of linear function, for instance, expression
```
typo canonical

romainbrenguier commented:
@tautschnig do you want to have a second look?
